### DISCUSSION TITLE
Performance issue with the Apollo extension and how to fix it

### DISCUSSION THREAD

**Discussion by hehaijin** [apollographql/apollo-client-devtools#1337] [Ideas]
I work on a relatively large project, the schema size for IntrospectionQuery is > 10MB. The Apollo extension has always feels slow, and when using it would slow down whole computer. I am a fan of the interface but some dev in our team would avoid using it
because of the performance issue. 

Sometime ago the extension started crashing so I looked into the code. I just created a pull request to fix the crash. But the performance issue is still there after the memory leak is fixed.

When debugging and looking through the code I noticed that client data, queries, mutations, cache are copied from main page to extension every 0.5s. So imagine that a big size of data is transferred from main page to extension every second even if nothing is happening. And this seems the main cause for the performance hit. when lowering this frequency, memory consumption and page responsiveness is significantly improved. 

Wondering if the issue is on the team's radar and have plan to fix?

1. Simplest would be allow this frequency to be user customizable?
2. A better solution would be listen to http requests, and only copy the data after queries/mutations are fired. some debouncing would be nice to have here. I suppose the cache can only be updated after a http request, but correct me if I am wrong. 
3. Best way would be to allow the client to emit some events after change and the extension can listen to that, so only the change needs to be copied.  But I suppose that needs a very big change.

Any thoughts?

**Comment by jerelmiller** [comment]
Hey @hehaijin 👋 

These are great observations! It is absolutely on our radar to fix. To be honest, the devtools were in pretty bad shape for a period of time and had a LOT of complaints about it simply not working. We wanted to get some of the glaring issues out of the way in regards to how the message passing works in the devtools to try and alleviate some of the major problems. So far these seem to have helped reduce a lot of the noise in that regard.

That being said, I think we are at a point where we can comfortably move onto other things. You are absolutely right on the fixes listed. Here are some thoughts on each of those points:

> A better solution would be listen to http requests, and only copy the data after queries/mutations are fired. some debouncing would be nice to have here. I suppose the cache can only be updated after a http request, but correct me if I am wrong.

You're absolutely right! We actually do have something called [`__actionHookForDevtools`](https://github.com/apollographql/apollo-client/blob/7f53db91e28fbde02cbe2c20dc91e4b40f9d6888/src/core/ApolloClient.ts#L569-L571) meant for this purpose, but I discovered that when this fired, it sent an event into devtools but nothing was listening for it. It just ended up slowing down the client for no reason, so I temporarily disabled it. In fact, you can see my comment about that here: https://github.com/apollographql/apollo-client-devtools/blob/6df9b7d2a1e4236eeabbc2e39f9db91da5e77193/src/extension/tab/hook.ts#L274-L283

Now that we've done a lot of rework on the message passing aspect, I'd like to explore reintroducing this client hook and see if this eliminates the need for polling. I'm really hoping so 🤞.

> Simplest would be allow this frequency to be user customizable?

I agree! I want to explore reintroducing that `__actionHookForDevtools` hook to get an idea of what triggers it first, but before that, something I think we can do in the even shorter term is swap out the `setInterval` with a `setTimeout` that triggers after the last event was successful. I think this might at least help somewhat in case copying the structure over takes > 0.5s and would mean at least we aren't completely using up your resources as often. Will this fix everything? Absolutely not, but might at least be a cheap short term solution for now.

If all else fails, I do like the idea of adding a user customizable frequency. Ideally we can avoid this, so this is my last resort if all else fails.

Thanks so much for opening the discussion!

**Comment by hehaijin** [comment]
Thanks @jerelmiller  for the detailed response. And very glad that you guys are working on the issue.
